ktor-http/common/src/io/ktor/http/URLParser.kt

```diff
@@ -41,6 +41,30 @@ internal fun URLBuilder.takeFromUnsafe(urlString: String): URLBuilder {
     val slashCount = count(urlString, startIndex, endIndex, '/')
     startIndex += slashCount
 
+    if (protocol.name == "file") {
+        require(slashCount == 3)
+        host = urlString.substring(startIndex, endIndex)
+        return this
+    }
+
+    if (protocol.name == "mailto") {
+        require(slashCount == 0)
+
+        val delimiter = urlString.indexOf("@", startIndex);
+        val passwordIndex = urlString.indexOfColonInHostPort(startIndex, delimiter)
+        if (delimiter < endIndex) {
+            if (passwordIndex != -1) {
+                user = urlString.substring(startIndex, passwordIndex)
+                password = urlString.substring(passwordIndex + 1, delimiter)
+            } else {
+                user = urlString.substring(startIndex, delimiter)
```

Comment:
So if we have no `@` character, then we get an empty username and a host without a first letter, right? I am not sure if it is allowed to have no `user@` in email, do you?